4.applications/grit-informed-aggregation/1.generate-profiles/scripts/nbconverted/1.aggregate-single-cell.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Aggregate single cells into bulk profiles using grit information
+#
+# We use normalized, **single-cell profiles** from the Cell Health experiment and **single-cell grit scores** (calculated with respect to Chr2 cutting controls and using normalized, feature selected profiles).
+#
+# Here we aggregate profiles from the Cell Health experiments using several "grit-informed" methods:
+# 1. Standard median aggregation
+# 2. Weighted mean, weighting by raw grit scores
+# 3. Weighted mean, weighting by softmax-transformed grit scores
+# 4. Weighted mean, weighting by grit scores clipped to 0. This assigns a minimum grit score of 0 to any cell with grit scores < 0.
+
+# In[2]:
+
+
+import os
+import glob
+import gzip
+from pathlib import Path
+from datetime import datetime
+
+import numpy as np
+import pandas as pd
+from scipy.special import softmax
+
+from pycytominer import aggregate, get_na_columns
+from pycytominer.cyto_utils import infer_cp_features
+from cytominer_eval import evaluate
+from scripts.utils import calculate_weighted_agg
+
+
+# In[3]:
+
+
+def merge_metadata(cell_line, level3_profile):
+    # load metadata file from Cell Health  data
+    commit = "07e4b40c39dd27084be36fbef4d64c5654b2960f"
+    base_url = f"https://github.com/broadinstitute/cell-health/raw/{commit}"
+    url = f"{base_url}/1.generate-profiles/data/metadata/platemap/DEPENDENCIES1_{cell_line}.csv"
+    platemap = pd.read_csv(url, sep=",")
+    platemap.columns = ["Metadata_" + str(x) for x in platemap.columns]
+
+    # merge with the aggregated files
+    meta_df = pd.merge(
+        level3_profile,
+        platemap,
+        left_on="Metadata_Well",
+        right_on="Metadata_well_position",
+    )
+    # # reorder columns for metadata to be in front
+    meta_df = meta_df[
+        sorted(meta_df, key=lambda x: x not in meta_df.filter(like="Metadata").columns)
+    ]
+
+    return meta_df
+
+
+# ## Load level 2 data
+
+# In[4]:
+
+
+plate_dict = {
+    "ES2": ["SQ00014613", "SQ00014614", "SQ00014615"],
+    "A549": ["SQ00014610", "SQ00014611", "SQ00014612"],
+    "HCC44": ["SQ00014616", "SQ00014617", "SQ00014618"],
+}
+
+
+# In[21]:
+
+
+get_ipython().run_cell_magic(
```

Comment:
Can you remind me if we double checked this line:

```python
scprofiles_df = (pd.merge(scprofiles_df, scgrit_df, 
     left_on=['Metadata_cell_identity', 'Metadata_Plate', 'Metadata_pert_name'], 
             right_on=['Metadata_cell_identity', 'Metadata_plate', 'Metadata_group'])
    )
```

The merging on Metadata_cell_identity is what concerns me. Given that we create this column based on the index of sc profiles _per plate_ (and per plate is key) are we sure that this will automatically align with the `sc_grit` file?

If you've double checked this already, great! I am happy. If not, please do.

Reply:
Also, this %%time block contains a lot of commented code that looks like it is important.

Can you add a brief comment (like 1 sentence) to this file indicating how you use the commented code? Something like "I ran this code twice, once without the following lines commented out" is all this needs I think.

Reply:
Excellent point. I copied the processing lines (used to process individual plate files and create the `Metadata_cell_identity` column) in the notebook that generated the `sc_grit files`, so I assumed they would be the same but I haven't explicitly checked.